### ISSUE TITLE
ZOWE - Java gpf failures

### ISSUE THREAD

**Issue by jiflemon** [SDL-Hercules-390/hyperion#210]
I'm running into issues running ZOWE on z/OS on Hercules. I get Java gpf failures during ZOWE start-up/initialisation.

I don't know if this is a Hercules bug, or just somethings that's not supported that ZOWE does or something else.... but it would be good to know if anyone else has got ZOWE working on Hercules.

ZOWE 'almost' works - I can get the Servers running, logon and access the ZOWE desktop and use the 3270 'app'...but the MVS and USS Explorer 'apps' don't work and I get these errors:

```
JVMDUMP039I Processing dump event "gpf", detail "" at 2019/04/23 20:09:39 - please wait.           
thread_wrapper+0xdfd8a048 (0x0000000000000000)                                                     
CELQHROD+0xe05d2398 (0x0000000000000000)                                                           
---------------------------------------                                                            
JVMDUMP039I Processing dump event "gpf", detail "" at 2019/04/23 20:09:39 - please wait.           
ZJ9SYM1+0x835ffcf8 (0x0000000000000000)                                                            
runJavaThread+0x835f3e10 (0x0000000000000000)                                                      
javaProtectedThreadProc+0x83555808 (0x0000000000000000)                                            
omrsig_protect+0xdfceebc8 (0x0000000000000000)                                                     
javaThreadProc+0x83551ad8 (0x0000000000000000)                                                     
thread_wrapper+0xdfd8a048 (0x0000000000000000)                                                     
---------------------------------------                                                            
Unhandled exception                                                                                
Type=Segmentation error vmState=0x00000000                                                         
J9Generic_Signal_Number=00000004 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000035  
Handler1=000000500871E960 Handler2=0000005008650C80                                                
gpr0=0000005000000003 gpr1=000000008CCA64D0 gpr2=000000008FFFFFD8 gpr3=FFFFFFFFFCCA5AEA            
gpr4=0000000000000000 gpr5=00000000216A5A70 gpr6=000000008CCA6458 gpr7=000000000000146A            
gpr8=000000000000056A gpr9=000000008CCA5550 gpr10=000000008CCA6458 gpr11=0000000021308300          
gpr12=0000000020E34018 gpr13=0000000021656F00 gpr14=0000005019D85D94 gpr15=0000000000000048        
fpc=0008000000000000 psw0=0785340180000000 psw1=0000005019D85DC0 sp=00000051781FE1C0               
bea=0000005019D85DC4                                                                               
fpr0=403e000000000000 fpr1=406f000000000000 fpr2=0000000000000000 fpr3=0000000000000000                         
fpr4=c03e000000000000 fpr5=0000000000000000 fpr6=0000000000000000 fpr7=404541462d000000                         
fpr8=0000000000000000 fpr9=0000000000000000 fpr10=0000000000000000 fpr11=0000000000000000                       
fpr12=0000000000000000 fpr13=0000000000000000 fpr14=0000000000000000 fpr15=0000000000000000                     
                                                                                                                
Compiled_method=com/ibm/jit/crypto/JITFullHardwareDigest.z_digestFinal([BIIJ)V                                  
Target=2_90_20180915_397057 (z/OS 02.03.00)                                                                     
CPU=s390x (8 logical CPUs) (0x2f305e000 RAM)                                                                    
----------- Stack Backtrace -----------                                                                         
JVMDUMP032I JVM requested System dump using 'IZUSVR.JVM.ZOWESVR6.D190423.T200939.X&DS' in response to an event  
JVMDUMP039I Processing dump event "gpf", detail "" at 2019/04/23 20:09:40 - please wait.                        
IEATDUMP in progress with options SDATA=(LPA,GRSQ,LSQA,NUC,PSA,RGN,SQA,SUM,SWA,TRT)                             
protectedIntrospectBacktraceSymbols+0xdfcebce0 (0x0000000000000000)                                             
omrsig_protect+0xdfceebc8 (0x0000000000000000)                                                                  
omrintrospect_backtrace_symbols+0xdfcebdd0 (0x0000000000000000)                                                 
generateDiagnosticFiles+0x835cbd18 (0x0000000000000000)                                                         
omrsig_protect+0xdfceebc8 (0x0000000000000000)                                                                  
structuredSignalHandler+0x835caa68 (0x0000000000000000)                                                         
masterSynchSignalHandler+0xdfcf2968 (0x0000000000000000)                                                        
__zerro+0xe07c4260 (0x0000000000000000)                                                                         
__zerros+0xe07c4590 (0x0000000000000000)                                                                        
```

**Comment by Fish-Git** [comment]
@jiflemon 

**I'm sorry to have to do this, but I'm going to have to reject this bug report for the following reasons:**

1. On the face of it, it does not appear to be a Hercules issue but rather more of a ZOWE issue.
2. You failed to identify which _**specific version**_ of Hercules the failure occurs on (i.e. Hercules `'version'` command). Providing a log file (and configuration file too, if pertinent) is typically very helpful in this regard.
3. You failed to identify which version of z/OS you are using.
4. You failed to identify which version of ZOWE you are using.
5. You failed to provide detailed instructions on how to recreate the problem. (I myself know absolutely _nothing_ about ZOWE, and I doubt any of the other developers do either.)
6. The failure information that you did provide does not pertain to Hercules at all but rather is specific to ZOWE only (which as I explained none of us know anything about) and thus is of absolutely no help at all with identifying the actual problem with Hercules (if it is indeed a problem with Hercules at all, which we don't know).

As a result, I am going to close this issue as "Unknown" at this time until such time as you can please provide more detailed and convincing evidence that the problem is actually with Hercules and which part of Hercules, based on the evidence that you need to provide, is the most likely one that is causing the problem.

### I would suggest that you proceed as follows:

1. Contact ZOWE product support directly for assistance with identifying why ZOWE is crashing. Is it a missing z/Architecture facility? Is is a specific machine instruction that isn't being executed correctly? Is it an I/O error or incorrect data being returned by Hercules?  Etc. Basically we need much stronger evidence of a problem with Hercules. Simply saying "Hercules isn't working right!" doesn't give us enough to go on.

2.  Once you have completed step 1 identifying what part of Hercules is actually failing, then please re-open this issue and provide the other missing information listed further above: Hercules version, z/OS version, ZOWE version, how to recreate the problem, etc.

Thanks.

And again I apologize for (temporarily?) rejecting your problem report, but you haven't exactly given us much to go on!

Please re-open this issue once you have gathered more evidence/information as described above.

Thanks.

**Comment by Fish-Git** [comment]
@jiflemon 

Another thing I would suggest doing is posting a question on the main _**[Hercules Support Forum](http://groups.yahoo.com/group/hercules-390)**_ asking whether any other Hercules users have tried using ZOWE at all and whether they too are having any problems with it:

* https://groups.yahoo.com/neo/groups/hercules-390/info

The URLs for the main Hercules Emulator support forum (as well as several other more specific Hercules related support forums) are listed on our _**[Technical Support](https://sdl-hercules-390.github.io/html/hercsupp.html)**_ web page:

* https://sdl-hercules-390.github.io/html/hercsupp.html

Thanks.

**Comment by mfsysprog** [comment]
The Compiled_method=com/ibm/jit/crypto/JITFullHardwareDigest.z_digestFinal([BIIJ)V message seems to indicate that there was an error in the crypto hardware implementation. It seems this error can by bypassed by disabling the crypto hardware acceleration by appending the line -Dcom.ibm.jit.crypto.nohardwarecrypt=true \  to the java statements in api-mediation/scripts/api-mediation-start-*.sh scripts in the ZOWE installation directory.

**Comment by mfsysprog** [comment]
Alternatively, it might be possible to adjust the jit level as per
https://www.ibm.com/support/pages/jit-problem-determination-ibm-sdk-using-xjit
and
https://www.ibm.com/support/knowledgecenter/en/SSYKE2_8.0.0/com.ibm.java.vm.80.doc/docs/jit_pd_failing_method.html
but I have not tested it.

**Comment by mfsysprog** [comment]
Hmm, still got the same abend running with -Dcom.ibm.jit.crypto.nohardwarecrypt=true so I'm afraid that one doesn't actually fix anything here...

**Comment by mfsysprog** [comment]
Have tested excluding the mentioned classes from jit, but got similar abend in the uncompiled java bytecode (Java_com_ibm_jit_crypto_JITFullHardwareDigest_z_1klmd_1native+0x859f7688)
Adding -Dcom.ibm.jit.crypto.nohardwarecrypt=true together with large enough timeout values seems to be the only way to get it to work somehow, managed to get the JES explorer to work one time, but it is very unstable

**Comment by mfsysprog** [comment]
The -Dcom.ibm.jit.crypto.nohardwaredigest=true is the bypass here, not -Dcom.ibm.jit.crypto.nohardwarecrypt. That also is a good indication that it is indeed a hardware error that is causing this abend, likely in the CPACF code handling digest functions somewhere.
@Fish-Git : would this give enough info to do some additional tracing in the hercules code?
Ps. The same error occured in zosmf after I bypassed the issue in zowe, when the mvs-explorer and uss-explorer are calling zosmf to do the login processing it seems. I've solved the latter by creating a var/zosmf/configuration/local_override.cfg file with JVM_OPTIONS=-Dcom.ibm.jit.crypto.nohardwaredigest=true 
This leads to the following message appearing:
IZUG020W: The value -Dcom.ibm.jit.crypto.nohardwaredigest=true was found 
 WARNING: for property JVM_OPTIONS in a local override file.

**Comment by Fish-Git** [comment]
> @Fish-Git : would this give enough info to do some additional tracing in the hercules code?

I have no idea. Everything you've written so far is complete gibberish -- _to me_.

What you've written might make perfect sense to someone who is familiar with the technology you are dealing with. but to someone like _me_ it makes no sense at all. It's complete gibberish.

As far as tracing Hercules code is concerned, what "additional tracing" do you suggest I/we do?? Is there a specific instruction you suspect might not be executing properly? If so, which one? I have no idea what type of "tracing" you feel is warranted.

In short, I'm still completely lost!&nbsp;`:(`

**Comment by mfsysprog** [comment]
@Fish-Git In short, the option tells the java runtime not to use any of the hardware acceleration functions for digest functions. I'm guessing it is one of the dyncrypt.c things. Let's just leave it for now as docu on how to bypass the abend. It takes a long time for everything to startup to actually trigger the abend so finding out what the actually issue is might take a long time.

**Comment by jiflemon** [comment]
Very interesting, how have you increased the timeouts ?

**Comment by mfsysprog** [comment]
See https://github.com/zowe/api-layer/issues/282 
I think I used something like 
    -Djersey.config.client.readTimeout=120000 \
    -Deureka.server.peer-node-read-timeout-ms=12000 \
    -Dapiml.gateway.timeoutMillis=120000 \
in all 3 api-mediation scripts together with the -Xshareclasses and -DLatencyUtils.useActualTime=false options. Not sure if they are neccessary (or even functional) in all three scripts, but at least it seems to work that way, combined with the above mentioned change in zosmf
